### PR TITLE
feat: key updater domain

### PR DESCRIPTION
**Pull request by tlm**
This PR introduces a new service layer called `keyupdater`. The key updater is responsible for helping update machines in Juju with new authorised keys that should be propogated on to a Juju machine within a model.

The service provides methods for listing the current keys that should be present on a machine and for also watching authorised key updates for a machine.

Previously this information was contained within model config and it was the keyupdater facades responsability to pull apart model config and amalgamate this information with the static public keys of the controller. All of this logic has now been moved down into the service allowing the facade to just ask for the keys for a machine.

Tests have been constructed to make sure that we are retaining the expected buisness logic originally found in the facade.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [x] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

Unit tests are sufficient. Please check test coverage against https://github.com/juju/juju/blob/main/apiserver/facades/agent/keyupdater/authorisedkeys.go

PR #17602 shows how this service will be used in practice.

## Documentation changes

No applicable. We expect full compatibility with what we have now.

## Links

**Jira card:** JUJU-6097

### CODE REVIEW

File: domain/keyupdater/service/controllerkey.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+// retrieving the controllers ssh keys.
+func NewControllerKeyService(st ControllerKeyState) *ControllerKeyService {
+	return &ControllerKeyService{
+		st: st,
+	}
+}
+
+// ControllerKeys returns the juju controllers ssh public keys. If no keys are
+// defined for the controller a zero length slice is returned.
+func (s *ControllerKeyService) ControllerKeys(ctx context.Context) ([]string, error) {
```
Comment (SimonRichardson) [inline domain/keyupdater/service/controllerkey.go:36]: Maybe `ControllerAuthorizedKeys`?

File: domain/schema/model.go [hunk 2]
```diff
@@ -17,7 +17,11 @@ import (
 //go:generate go run ./../../generate/triggergen -db=model -destination=./model/triggers/objectstore-triggers.gen.go -package=triggers -tables=object_store_metadata_path
 //go:generate go run ./../../generate/triggergen -db=model -destination=./model/triggers/secret-triggers.gen.go -package=triggers -tables=secret_metadata,secret_rotation,secret_revision,secret_revision_expire,secret_revision_obsolete,secret_revision,secret_reference
 //go:generate go run ./../../generate/triggergen -db=model -destination=./model/triggers/network-triggers.gen.go -package=triggers -tables=subnet
+<<<<<<< HEAD
 //go:generate go run ./../../generate/triggergen -db=model -destination=./model/triggers/machine-triggers.gen.go -package=triggers -tables=machine
+=======
+//go:generate go run ./../../generate/triggergen -db=model -destination=./model/triggers/user-public-ssh-key.gen.go -package=triggers -tables=user_public_ssh_key
+>>>>>>> 5e8972342f (feat: Introduces service layer for key updater.)
```
Comment (SimonRichardson) [inline domain/schema/model.go:24]: Whoops.

File: domain/keyupdater/service/service.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+		watcherFactory: watcherFactory,
+	}
+}
+
+// AuthorisedKeysForMachine is responsible for fetching the authorised keys that
+// should be available on a machine. The following errors can be expected:
+// - [github.com/juju/errors.NotValid] if the machine id is not valid.
+// - [github.com/juju/juju/domain/machine/errors.NotFound] if the machine does
+// not exist.
+func (s *Service) AuthorisedKeysForMachine(
```
Comment (SimonRichardson) [inline domain/keyupdater/service/service.go:108]: If we've got `GetControllerKeys`, then we should prefix this with `Get` to be consistent.

**Review by SimonRichardson — COMMENTED** [review]
Looks good, just a few nips and tucks.

File: domain/schema/model.go (hunk 2 above)
Comment (tlm) [reply domain/schema/model.go]: Done & Done

File: domain/keyupdater/service/controllerkey.go (hunk 1 above)
Comment (tlm) [reply domain/keyupdater/service/controllerkey.go:36]: Done & Done

File: domain/keyupdater/service/service.go (hunk 3 above)
Comment (tlm) [reply domain/keyupdater/service/service.go:108]: Done & Done

File: domain/keyupdater/service/controllerkey.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/juju/juju/controller"
+	"github.com/juju/juju/internal/ssh"
+)
+
+// ControllerKeyService provides a service for interacting the the underlying
```
Comment (hpidcock) [inline domain/keyupdater/service/controllerkey.go:14]:
```suggestion
// ControllerKeyService provides a service for interacting with the underlying
```

File: domain/keyupdater/service/controllerkey.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+// ssh keys stored for a controller.
+type ControllerKeyService struct {
+	st ControllerKeyState
+}
+
+// ControllerKeyState defines the state layer for interacting with a controllers
```
Comment (hpidcock) [inline domain/keyupdater/service/controllerkey.go:20]:
```suggestion
// ControllerKeyState defines the state layer for interacting with a controller's
```

File: domain/keyupdater/service/controllerkey.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+// config and ssh keys.
+type ControllerKeyState interface {
+	GetControllerConfigKeys(context.Context, []string) (map[string]string, error)
+}
+
+// NewControllerKeyService constructs a new [ControllerKeyService] for
+// retrieving the controllers ssh keys.
```
Comment (hpidcock) [inline domain/keyupdater/service/controllerkey.go:27]:
```suggestion
// retrieving the controller's ssh keys.
```

File: domain/keyupdater/service/controllerkey.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+func NewControllerKeyService(st ControllerKeyState) *ControllerKeyService {
+	return &ControllerKeyService{
+		st: st,
+	}
+}
+
+// ControllerAuthorisedKeys returns the juju controllers ssh public keys. If no
```
Comment (hpidcock) [inline domain/keyupdater/service/controllerkey.go:34]:
```suggestion
// ControllerAuthorisedKeys returns the juju controller's ssh public keys. If no
```

File: domain/keyupdater/service/controllerkey.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+// keys are defined for the controller a zero length slice is returned.
```
Comment (hpidcock) [inline domain/keyupdater/service/controllerkey.go:35]:
```suggestion
// keys are defined for the controller, a zero length slice is returned.
```

File: internal/ssh/authorisedkey.go [hunk 9]
```diff
@@ -39,3 +40,20 @@ func ParsePublicKey(key string) (PublicKey, error) {
 		Comment: comment,
 	}, nil
 }
+
+// SplitAuthorisedKeys extracts a key slice from the specified key data,
+// by splitting the key data into lines and ignoring comments and blank lines.
+func SplitAuthorisedKeys(keyData string) []string {
```
Comment (hpidcock) [inline internal/ssh/authorisedkey.go:46]: I would argue, that here, it is AuthorizedKeys with a `z`. The Americanz won on thiz one.

File: domain/schema/model/triggers/user-public-ssh-key.gen.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import (
+	"fmt"
+
+	"github.com/juju/juju/core/database/schema"
+)
+
+
+// ChangeLogTriggersForUserPublicSshKey generates the triggers for the 
+// user_public_ssh_key table.
+func ChangeLogTriggersForUserPublicSshKey(columnName string, namespaceID int) func() schema.Patch {
```
Comment (hpidcock) [inline domain/schema/model/triggers/user-public-ssh-key.gen.go:14]:
```suggestion
func ChangeLogTriggersForUserPublicSSHKey(columnName string, namespaceID int) func() schema.Patch {
```

It is not the onomatopoeic word. Can you possible fix the code generator to correct this?

File: internal/ssh/authorisedkey.go (hunk 9 above)
Comment (hpidcock) [reply internal/ssh/authorisedkey.go:46]: Specifically, when we refer to SSH Authorized Keys. This is in the SSH package, it is parsing the authorized_keys form of keys that are authorised.

**Review by hpidcock — APPROVED** [review]
LGTM, I only have comments, but you asked for a green tick, I still need to review it.

File: domain/keyupdater/service/controllerkey.go (hunk 4 above)
Comment (tlm) [reply domain/keyupdater/service/controllerkey.go:14]: Done & Done

File: domain/keyupdater/service/controllerkey.go (hunk 5 above)
Comment (tlm) [reply domain/keyupdater/service/controllerkey.go:20]: Done & Done

File: domain/keyupdater/service/controllerkey.go (hunk 6 above)
Comment (tlm) [reply domain/keyupdater/service/controllerkey.go:27]: Done & Done

File: domain/keyupdater/service/controllerkey.go (hunk 7 above)
Comment (tlm) [reply domain/keyupdater/service/controllerkey.go:34]: Done & Done

File: domain/keyupdater/service/controllerkey.go (hunk 8 above)
Comment (tlm) [reply domain/keyupdater/service/controllerkey.go:35]: Done & Done

File: internal/ssh/authorisedkey.go (hunk 9 above)
Comment (tlm) [reply internal/ssh/authorisedkey.go:46]: Done & Done

File: domain/schema/model/triggers/user-public-ssh-key.gen.go (hunk 10 above)
Comment (tlm) [reply domain/schema/model/triggers/user-public-ssh-key.gen.go:14]: I don't think we can.

**Pull request merged**